programs/host/frontend.c

```diff
@@ -0,0 +1,937 @@
+ *
+ *  See library/serialize.c for a description of the serialization format.
+ *
+ *  Copyright (C) 2018, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#include "frontend-config.h"
+
+#include <stdint.h>
+#include <errno.h>
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#define mbedtls_calloc    calloc
+#define mbedtls_free      free
+#define mbedtls_fprintf   fprintf
+#include <unistd.h>
+#include <dirent.h>
+#include <sys/stat.h>
+#endif
+
+FILE * fdbg = NULL;
+
+static int debug_verbose = 0;
+#define DBG( fmt, ... ) do {                                                \
+               fprintf( fdbg, "%s:%d:%s: " fmt "\n",                  \
+                        __FILE__, __LINE__, __FUNCTION__, ##__VA_ARGS__ );  \
+               fflush( fdbg ); \
+} while (0)
+
+#include "mbedtls/serialize.h"
+#include "mbedtls/net_sockets.h"
+#include "mbedtls/fsio.h"
+
+/** State of the offloading frontend. */
+typedef enum {
+    /** The communication channel is broken */
+    MBEDTLS_SERIALIZE_STATUS_DEAD = 0,
+    /** All conditions nominal */
+    MBEDTLS_SERIALIZE_STATUS_OK = 1,
+    /** Out of memory for a function's parameters.
+     * Normal operation can resume after the next stack flush. */
+    MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY = 2,
+    /** An exit command has been received */
+    MBEDTLS_SERIALIZE_STATUS_EXITED = 3,
+} mbedtls_serialize_status_t;
+
+/** An input or output to a serialized function.
+ *
+ * Inputs are kept as a linked list, outputs are kept in an array. There's
+ * no deep reason for that, in retrospect an array would do fine for inputs. */
+struct mbedtls_serialize_item;
+typedef struct mbedtls_serialize_item {
+    struct mbedtls_serialize_item *next; /**< Next input */
+    size_t size; /**< Size of the following data in bytes */
+    /* followed by the actual data */
+} mbedtls_serialize_item_t;
+
+/** Specialization of mbedtls_serialize_item_t with enough room for
+ * data for a uint32. */
+typedef struct {
+    mbedtls_serialize_item_t meta;
+    unsigned char data[4];
+} mbedtls_serialize_uint32_t;
+
+
+static void *item_buffer( mbedtls_serialize_item_t *item )
+{
+    return( (unsigned char*) item + sizeof( *item ) );
```

Comment:
This makes assumptions on the layout of structures specializing `mbedtls_serialize_item_t` - is it portable across platforms? For example, on a system with 64-bit pointers but 32-bit `size_t`, `mbedtls_serialize_item_t meta;` would be padded with 32 bits within `mbedtls_serialize_uint32_t`, wouldn't it?

Reply:
Shouldn't `+ sizeof( *item )` skip over the structure memory irrespective of layout?

Reply:
If you had something like
```
typedef struct
{
    uint8_t a;
    uint16_t b;
} A;

typedef struct
{
    A a;
    uint64_t c;
} B;
```
then `sizeof( A ) == 4` but the offset of `c` in `B` is `8`. So there's padding for `A` alone, and for `A` when used within `B`.

However, there is no such additional padding of `mbedtls_serialize_item_t` when used within `mbedtls_serialize_uint32_t`, because there is no padding constraint on `uint8_t[4]`. So I think we can disregard my comment.

Reply:
I see what you mean. But as you said there are no padding constraints on `uint8_t` fields/array.
https://github.com/ARMmbed/mbedtls/blob/3a630be92b7dd13cd033aafbf71d199aafca3ffe/programs/host/frontend.c#L91
AFAIK if it was any other type than byte size then the alignment would have been an issue.